File: libs/dynamo-repo/src/system-config-repo.ts
import { GetCommand } from "@aws-sdk/lib-dynamodb";
import {
  systemConfigKey,
  fromSystemConfigItem,
  DEFAULT_SYSTEM_CONFIG,
} from "@bifrost/dynamo-schema";
import { getDocClient, getTableName } from "./client.js";

export interface SystemConfig {
  readonly heartbeatIntervalSeconds: number;
  readonly staleThresholdSeconds: number;
  readonly syncTimeoutSeconds: number;
  readonly maxRetries: number;
  readonly driftCheckIntervalSeconds: number;
  readonly autoPromoteEnabled: boolean;
  readonly autoPromoteStaleSeconds: number;
}

export async function getSystemConfig(): Promise<SystemConfig> {
  const result = await getDocClient().send(
    new GetCommand({
      TableName: getTableName(),
      Key: systemConfigKey(),
    }),
  );

  if (!result.Item) {
    return { ...DEFAULT_SYSTEM_CONFIG };
  }

  const item = fromSystemConfigItem(result.Item);
  return {
    heartbeatIntervalSeconds:
      item.heartbeatIntervalSeconds ?? DEFAULT_SYSTEM_CONFIG.heartbeatIntervalSeconds,
    staleThresholdSeconds:
      item.staleThresholdSeconds ?? DEFAULT_SYSTEM_CONFIG.staleThresholdSeconds,
    syncTimeoutSeconds:
      item.syncTimeoutSeconds ?? DEFAULT_SYSTEM_CONFIG.syncTimeoutSeconds,
    maxRetries: item.maxRetries ?? DEFAULT_SYSTEM_CONFIG.maxRetries,
    driftCheckIntervalSeconds:
      item.driftCheckIntervalSeconds ?? DEFAULT_SYSTEM_CONFIG.driftCheckIntervalSeconds,
    autoPromoteEnabled:
      item.autoPromoteEnabled ?? DEFAULT_SYSTEM_CONFIG.autoPromoteEnabled,
    autoPromoteStaleSeconds:
      item.autoPromoteStaleSeconds ?? DEFAULT_SYSTEM_CONFIG.autoPromoteStaleSeconds,
  };
}

File: apps/spark-server/src/repo/system_config_repo.rs
use crate::domain::SystemConfig;
use crate::error::AppResult;
use sqlx::SqlitePool;

impl Default for SystemConfig {
    fn default() -> Self {
        SystemConfig {
            heartbeat_interval_seconds: 30,
            stale_threshold_seconds: 120,
            sync_timeout_seconds: 60,
            max_retries: 10,
            drift_check_interval_seconds: 300,
            auto_promote_enabled: true,
            auto_promote_stale_seconds: 120,
        }
    }
}

pub async fn get_system_config(pool: &SqlitePool) -> AppResult<SystemConfig> {
    let cfg = sqlx::query_as::<_, SystemConfig>(
        "SELECT heartbeat_interval_seconds, stale_threshold_seconds, sync_timeout_seconds, \
         max_retries, drift_check_interval_seconds, auto_promote_enabled, auto_promote_stale_seconds \
         FROM system_config WHERE id = 1",
    )
    .fetch_optional(pool)
    .await?
    .unwrap_or_default();
    Ok(cfg)
}
